### DISCUSSION TITLE
authStore not being valid after successfull OAuth2 login

### DISCUSSION THREAD

**Discussion by Phillor** [pocketbase/pocketbase#4190] [Q&A]
Hi there,

so I'm working with SvelteKit CSR with the adapter-static. I was updating a project from last year where I was using the old fashion way of authenticating a user via OAuth2 (manual code exchange). This was working fine and without any problems.

Because of the simplicity of the new variant I decided to switch over to the All in one solution which works fine as well. So I thought.

My problem ist that right after login (which succeeds) my app checks if the authStore is valid in a separate Javascript module which is responsible for protecting routes clientside. This also succeeds and I'm able to enter the page that should be only accessible by authenticated users.

Now right after login I have 4 SvelteKit Stores that are also checking if the authStore is valid (just to avoid errors in the clients console) and then subscribing to one collection each on pocketbase. There the authStore check fails after the first login so the page is completely empty (thus the stores are not fetching any data because authStore invalid). When I reload the page everything works again.

Also when I logout and login again everything works fine too. Only when I login inside a new browser tab this problems appears again until I refresh the page.

I find it a bit confusing because my routeguard is actually able to verify the authStore but my stores are not. This is only after switching to the OAuth2 All in one solution. Do you guys have an idea what this could be?

Thank you in advance!

**Comment by ganigeorgiev** [comment]
I don't think I'll be able to help you with this without a **minimal** code sample (especially those _"4 SvelteKit Stores that are also checking if the authStore is valid"_) but note that `pb.authStore` is not "reactive" like the Svelte variables so if you are expecting your js code to update on `pb.authStore` change you'll have to listen to the [`pb.authStore.onChange((token, model) => { ... })`](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#auth-store) method and trigger the Svelte stores update from there.

**Reply by Phillor** [reply]
You are absolutely right. I'm sorry for that!
I'm aware of pb.authStore not being reactive. I was just wondering why my protectRoutes function is actually able to verify the authStore:
```
import { browser } from '$app/environment';
import { goto } from '$app/navigation';
import { pb } from '$lib/pocketbase';

export function protectRoutes(route:string) {
    if (pb.authStore.isValid){
        if (!pb.authStore.model?.profile || pb.authStore.model?.expand.profile.locked){
            if (browser){
                goto('/locked', { replaceState: true });
            }
        }else if (route.startsWith('/login') || route.startsWith('/locked')){
            if (browser){
                goto('/', { replaceState: true});
            }
        }else if (route.startsWith('/admin')){
            if (!pb.authStore.model?.admin){
                if (browser){
                    goto('/', { replaceState: true});
                }
            }
        }
    }else{
        if (!route.startsWith('/login')){
            if (browser){
                goto('/login', { replaceState: true});
            }
        }
    }
}
```

While my Stores are not. Here is an example of one of the stores. They are all build the same, just different collections:

```
import { writable, get } from 'svelte/store';
import { pb } from '$lib/pocketbase';

export const unitStore = writable([] as any);

if (pb.authStore.isValid){
    try {
        const users = await pb.collection('units').getFullList(undefined, {
            sort: 'name',
        });
        unitStore.set(users);
    } catch (error) {
        console.log('Error:', error);
    }
}

if (pb.authStore.isValid){
    $: pb.collection('units').subscribe("*", function (e) {
        const changedUnit = e.record;
        let existingUnits:any = get(unitStore);
        if (e.action === "update"){
            for (const unit in existingUnits) {
                if (existingUnits[unit].id === changedUnit.id){
                    existingUnits[unit].name = changedUnit.name;
                    existingUnits[unit].status = changedUnit.status;
                    existingUnits[unit].district = changedUnit.district;
                }
            }
        } else if(e.action === "create"){
            existingUnits.push(changedUnit);
        } else if(e.action === "delete"){
            for (const unit in existingUnits) {
                if (existingUnits[unit].id === changedUnit.id){
                    existingUnits.splice(unit, 1);
                }
            }
        }
        
        unitStore.set(existingUnits);
    });
}
```

Both are used just after the OAuth2 login. protectRoutes validation is working. The one of the store is not. It seems like the validation of the store is just happening to early when the authStore is not populated yet because after a reload it just works.

Here is the call of both in the +page.svelte file:

```
import Unit from '$lib/components/Unit.svelte';
 import { profileStore } from '$lib/stores/profileStore';
 import { unitStore } from '$lib/stores/unitStore';
 import { districtStore } from '$lib/stores/districtStore';
 import { statusCodeStore } from '$lib/stores/statusCodeStore';
 import { pb } from '$lib/pocketbase';

 import { page } from '$app/stores';
 import { protectRoutes } from '$lib/routeguard';

 protectRoutes($page.url.pathname);
    
	

 let officerList:any;
 let unitList:any;
 let districtList:any;
 let statusCodeList:any;

 $: unitStore.subscribe(value => {
     unitList = value;
 })

 $: profileStore.subscribe(value => {
     officerList = value;
 });

 $: districtStore.subscribe(value => {
     districtList = value;
 });

 $: statusCodeStore.subscribe(value => {
     statusCodeList = value;
 })
```

**Reply by ganigeorgiev** [reply]
I have no idea why it doesn't work, but the above code doesn't look correct to me. For example I don't understand the point of having `$:` statement for the `pb.collection('units').subscribe` (actually I'm not even sure whether it is ok to have a wrapping `if` block for the `$:` statement in the first place and whether it will be treated just as regular JS label or Svelte reactive declaration).

**Reply by ganigeorgiev** [reply]
The easiest way to inspect the firing order of the checks is to add debug pointers or simple `console.log`s right after the OAuth2 async call and similarly in your components.

**Reply by Phillor** [reply]
Alright. I will take a look at the things you mentioned. Thank you very much! Very kind of you that you took the time to help me out.

**Comment by jonshipman** [comment]
Checkout my sveltekit boilerplate here: https://github.com/jonshipman/sveltekit-pb-boilerplate

I'm not using adapter static, but you can easily modify it to fit your use case. But you should be using advanced layout grouping (https://kit.svelte.dev/docs/advanced-routing#advanced-layouts) instead of that `protectRoutes` function checking. Throw all your protected routes into a `(protected)` folder and check in a +layout.ts file.
